### PR TITLE
Return from main() invokes exit_current_task()

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by beezow** [comment]
What is the motivation for this change?

File: sys/userlib/src/arch/arm_m.rs [hunk 1]
```diff
@@ -65,11 +65,14 @@ pub unsafe extern "C" fn _start() -> ! {
                 @ a sym operand because it's a Rust func and may be mangled.
                 bl {main}
 
+                bl {exit_current_task}
+
                 @ The noreturn option below will automatically generate an
                 @ undefined instruction trap past this point, should main
```
Comment (furquan-rivos) [inline sys/userlib/src/arch/arm_m.rs]: comment needs update

File: sys/userlib/src/arch/arm_m.rs [hunk 2]
```diff
@@ -65,11 +65,14 @@ pub unsafe extern "C" fn _start() -> ! {
                 @ a sym operand because it's a Rust func and may be mangled.
```
Comment (furquan-rivos) [inline sys/userlib/src/arch/arm_m.rs:60]: Comment above needs update

File: sys/userlib/src/arch/arm_m.rs [hunk 3]
```diff
@@ -65,11 +65,14 @@ pub unsafe extern "C" fn _start() -> ! {
...
                 bl {main}
```
Comment (furquan-rivos) [inline sys/userlib/src/arch/arm_m.rs]: What do you think about adding a userlib common entry point: `bl userlib_main` which does the following:
```
#[no_mangle]
pub fn userlib_main() {
    main();
    kipc::exit_current_task();
}
```

Then you won't have to repeat the same in different architectures. And have the logic commented in one place as well.

File: task/one-shot/src/main.rs [hunk 4]
```diff
@@ -8,10 +8,7 @@
 use userlib::*;
 
 #[export_name = "main"]
-fn main() -> ! {
+fn main() {
     sys_log!("Hello world from one-shot task!");
     sys_log!("Exiting now!");
-    kipc::exit_current_task();
```
Comment (furquan-rivos) [inline task/one-shot/src/main.rs:14]: Nice!! I like this :smile: It feels more natural than to add a `kipc::exit_current_task()` and `unreachable!()` in every one-shot task!!

File: sys/userlib/src/arch/arm_m.rs (hunk 3 above)
Comment (ghost) [reply sys/userlib/src/arch/arm_m.rs]: That's a much better idea than all the arch-specific changes. I'll update the PR shortly.

**Comment by ghost** [comment]
> What is the motivation for this change?

Commit message has been update to better explain the motivation.

**Review by ghost — DISMISSED** [review]
Good stuff - one fewer foot-gun.

**Review by FawazTirmizi — DISMISSED** [review]

File: sys/userlib/src/lib.rs [hunk 5]
```diff
@@ -644,3 +644,17 @@ pub fn sys_reply_fault(task_id: TaskId, reason: ReplyFaultReason) {
...
+
+// Provided by the user program:
+extern "Rust" {
+    fn main();
+}
+
+/// Stub to call the user program's main() function and then exit
+/// cleanly should it return.
+pub fn userlib_main() -> ! {
+    unsafe {
```
Comment (furquan-rivos) [inline sys/userlib/src/lib.rs:661]: SAFETY comment?

Comment (ghost) [reply sys/userlib/src/lib.rs:661]: Added

**Review by furquan-rivos — APPROVED** [review]

**Pull request merged**